feat: Collector Metric Exporter[2/x] Create CollectorMetricExporterBase

davidwitten opened the pull request:
# Summary
This PR addresses Issue #1109.

Currently, the Collector Exporter only sends traces to the collector. In Python, Java, and Go, for example, metric exporting already exists for the collector. This PR adds that feature. I created two versions, one for Node and the other for browser. The node version uses gRPC, and the browser uses Beacon or XHR. It's up to date with all previous collector exporter commits.

# What this PR does

This is the 2nd part of this long [PR](https://github.com/open-telemetry/opentelemetry-js/pull/1248).

This PR creates `CollectorMetricExporterBase`, modeled off `CollectorTraceExporterBase`. I also create unit tests and add two mock metrics to test with, a mock counter and a mock observer. The base class does nothing, so this won't break anything.

## Next Steps
The next diffs will be as follows 
1. Add transform functions and tests
2. E2E Metric exporter for Browser with tests
3. E2E Metric exporter for Node with tests

davidwitten commented:
This PR: https://github.com/open-telemetry/opentelemetry-js/pull/1254 reverted my proto change, so I'm doing it again.

mayurkale22 commented on packages/opentelemetry-exporter-collector/src/CollectorMetricExporterBase.ts:
```diff
@@ -0,0 +1,122 @@
+      })
+      .catch(
+        (
+          error: collectorTypes.opentelemetryProto.collector.trace.v1.ExportTraceServiceError
```
This looks fishy, why are we using `ExportTraceServiceError` for metric?

davidwitten replied:
Ah yeah, that type isn't actually specified in the proto. It's just a template for errors. So, I'm renaming it to `ExportServiceError` and moving it out of the proto

mayurkale22 commented on packages/opentelemetry-exporter-collector/src/CollectorMetricExporterBase.ts:
```diff
@@ -0,0 +1,122 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { MetricExporter, MetricRecord } from '@opentelemetry/metrics';
```
Minor: please include empty line after the license block.

naseemkullah commented on packages/opentelemetry-exporter-collector/src/CollectorMetricExporterBase.ts:
```diff
@@ -0,0 +1,119 @@
+    this.serviceName = config.serviceName || DEFAULT_SERVICE_NAME;
+    this.url = this.getDefaultUrl(config.url);
+    this.attributes = config.attributes;
+    if (typeof config.hostName === 'string') {
+      this.hostName = config.hostName;
```
```suggestion
    if (typeof config.hostname === 'string') {
      this.hostname = config.hostname;
```

nit: hostname is one word

naseemkullah commented on packages/opentelemetry-exporter-collector/src/CollectorMetricExporterBase.ts:
```diff
@@ -0,0 +1,119 @@
+  public readonly serviceName: string;
+  public readonly url: string;
+  public readonly logger: Logger;
+  public readonly hostName: string | undefined;
```
```suggestion
  public readonly hostname: string | undefined;
```

naseemkullah commented on packages/opentelemetry-exporter-collector/test/common/CollectorMetricExporter.test.ts:
```diff
@@ -0,0 +1,176 @@
+    beforeEach(() => {
+      onInitSpy = sinon.stub(CollectorMetricExporter.prototype, 'onInit');
+      collectorExporterConfig = {
+        hostName: 'foo',
```
```suggestion
        hostname: 'foo',
```

naseemkullah commented on packages/opentelemetry-exporter-collector/test/common/CollectorMetricExporter.test.ts:
```diff
@@ -0,0 +1,176 @@
+    });
+
+    describe('when config contains certain params', () => {
+      it('should set hostName', () => {
+        assert.strictEqual(collectorExporter.hostName, 'foo');
```
```suggestion
      it('should set hostname', () => {
        assert.strictEqual(collectorExporter.hostname, 'foo');
```

naseemkullah commented on packages/opentelemetry-exporter-collector/test/common/CollectorMetricExporter.test.ts:
```diff
@@ -0,0 +1,176 @@
+        'onShutdown'
+      );
+      collectorExporterConfig = {
+        hostName: 'foo',
```
```suggestion
        hostname: 'foo',
```

naseemkullah commented on packages/opentelemetry-exporter-collector/test/common/CollectorMetricExporter.test.ts:
```diff
@@ -0,0 +1,176 @@
+    });
+
+    describe('when config contains certain params', () => {
+      it('should set hostName', () => {
```
```suggestion
      it('should set hostname', () => {
```

davidwitten commented:
@naseemkullah Actually, I realized that `this.hostname` is also set in CollectorTraceExporterBase and is never used. Is it okay to just remove it from both?

naseemkullah commented:
> @naseemkullah Actually, I realized that `this.hostname` is also set in CollectorTraceExporterBase and is never used. Is it okay to just remove it from both?

At a quick glance, imho, we can remove `hostname` but we should add `host` and `port`, which should be used with Node, whereas `url` with browser, since for reasons beyond my understanding they must be configured differently. @open-telemetry/javascript-maintainers please advise.

naseemkullah commented:
All that to say I'm ok with removing `hostname` and (eventually adding `host` and `port`, probably not in this PR,  if Node cannot be configured via `url` with a url).

davidwitten commented:
Thanks @naseemkullah! For now, I just changed `hostName` to `hostname` and added more unit tests. I'll make another PR to remove it and replace it with host/port.
